src/textual/validation.py

```diff
@@ -0,0 +1,505 @@
+
+    @staticmethod
+    def merge(results: Sequence["ValidationResult"]) -> "ValidationResult":
+        """Merge multiple ValidationResult objects into one.
+
+        Args:
+            results: List of ValidationResult objects to merge.
+
+        Returns:
+            Merged ValidationResult object.
+        """
+        is_valid = all(result.is_valid for result in results)
+        failures = [failure for result in results for failure in result.failures]
+        if is_valid:
+            return ValidationResult.success()
+        else:
+            return ValidationResult.failure(failures)
+
+    @staticmethod
+    def success() -> ValidationResult:
+        """Construct a successful ValidationResult.
+
+        Returns:
+            A successful ValidationResult.
+        """
+        return ValidationResult()
+
+    @staticmethod
+    def failure(failures: Sequence[Failure]) -> ValidationResult:
+        """Construct a failure ValidationResult.
+
+        Args:
+            failures: The failures.
+
+        Returns:
+            A failure ValidationResult.
+        """
+        return ValidationResult(failures)
+
+    @property
+    def failure_descriptions(self) -> list[str]:
+        """Utility for extracting failure descriptions as strings.
+
+        Useful if you don't care about the additional metadata included in the `Failure` objects.
+
+        Returns:
+            A list of the string descriptions explaining the failing validations.
+        """
+        return [
+            failure.description
+            for failure in self.failures
+            if failure.description is not None
+        ]
+
+    @property
+    def is_valid(self) -> bool:
+        return len(self.failures) == 0
+
+
+@dataclass
+class Failure:
+    """Information about a validation failure."""
+
+    validator: Validator
+    """The Validator which produced the failure."""
+    value: str | None = None
+    """The value which resulted in validation failing."""
+    description: str | None = None
+    """An optional override for describing this failure. Takes precedence over any messages set in the Validator."""
+
+    def __post_init__(self) -> None:
+        # If a failure message isn't supplied, try to get it from the Validator.
+        if self.description is None and self.validator is not None:
+            if self.validator.failure_description is not None:
+                self.description = self.validator.failure_description
+            else:
+                self.description = self.validator.describe_failure(self)
+
+    def __rich_repr__(self) -> rich.repr.Result:  # pragma: no cover
+        yield self.value
+        yield self.validator
+        yield self.description
+
+
+class Validator(ABC):
+    """Base class for the validation of string values.
+
+    Commonly used in conjunction with the `Input` widget, which accepts a
+    list of validators via its constructor. This validation framework can also be used to validate any 'stringly-typed'
```

Comment:
`stringly-typed` ? Is this an established term?

Reply:
It's a pretty well-known term referring to data that shouldn't be a string but is a string (for whatever reason). e.g. `"10"` is a stringly-typed integer.

Reply:
Fair enough. New one on me!